File: TOOL-LOCAL/Web/src/features/vietsub/vietsubSubtitleStyle.ts
import type { CSSProperties } from 'react';
import type { VietsubSubtitleStyle, VietsubSubtitleStylePreset } from './types';

export const defaultVietsubSubtitleStyle: VietsubSubtitleStyle = {
  presetId: 'READABLE',
  fontFamily: 'Arial',
  fontSizePercent: 4.4,
  bold: true,
  italic: false,
  textColor: '#FFFFFF',
  textOpacity: 1,
  outlineColor: '#000000',
  outlineOpacity: 0.95,
  outlineWidthPercent: 0.18,
  shadowColor: '#000000',
  shadowOpacity: 0.65,
  shadowOffsetPercent: 0.16,
  backgroundEnabled: true,
  backgroundColor: '#04080D',
  backgroundOpacity: 0.58,
  alignment: 'BOTTOM_CENTER',
  verticalPosition: 'BOTTOM',
  positionXPercent: 50,
  positionYPercent: 93,
  bottomMarginPercent: 7,
  horizontalMarginPercent: 6,
  maxWidthPercent: 88,
  lineHeight: 1.25,
  maxLines: 2
};

export const vietsubSubtitlePresets: ReadonlyArray<{
  id: Exclude<VietsubSubtitleStylePreset, 'CUSTOM'>;
  label: string;
  description: string;
  style: VietsubSubtitleStyle;
}> = [
  {
    id: 'READABLE',
    label: 'Dễ đọc',
    description: 'Nền tối nhẹ, phù hợp đa số video.',
    style: defaultVietsubSubtitleStyle
  },
  {
    id: 'OUTLINE',
    label: 'Viền rõ',
    description: 'Chữ trắng, viền đen rõ và không dùng hộp nền.',
    style: {
      ...defaultVietsubSubtitleStyle,
      presetId: 'OUTLINE',
      fontFamily: 'Tahoma',
      backgroundEnabled: false,
      outlineWidthPercent: 0.34,
      shadowOffsetPercent: 0.12
    }
  },
  {
    id: 'TIKTOK',
    label: 'TikTok',
    description: 'Chữ lớn, đậm và nổi bật cho video ngắn.',
    style: {
      ...defaultVietsubSubtitleStyle,
      presetId: 'TIKTOK',
      fontFamily: 'Verdana',
      fontSizePercent: 5.7,
      backgroundEnabled: false,
      outlineWidthPercent: 0.42,
      shadowOffsetPercent: 0.2,
      verticalPosition: 'BOTTOM',
      bottomMarginPercent: 12,
      positionYPercent: 88,
      maxWidthPercent: 82,
      horizontalMarginPercent: 9,
      maxLines: 2
    }
  },
  {
    id: 'YELLOW',
    label: 'Vàng nổi bật',
    description: 'Chữ vàng, viền đen dễ đọc trên khung hình sáng.',
    style: {
      ...defaultVietsubSubtitleStyle,
      presetId: 'YELLOW',
      fontFamily: 'Tahoma',
      textColor: '#FFE45C',
      backgroundEnabled: false,
      outlineWidthPercent: 0.32,
      shadowOffsetPercent: 0.15
    }
  },
  {
    id: 'VERTICAL',
    label: 'Video dọc',
    description: 'Chữ lớn hơn và lề dưới an toàn cho video ngắn.',
    style: {
      ...defaultVietsubSubtitleStyle,
      presetId: 'VERTICAL',
      fontSizePercent: 5.2,
      bottomMarginPercent: 10,
      positionYPercent: 90,
      horizontalMarginPercent: 8,
      maxWidthPercent: 84
    }
  },
  {
    id: 'CINEMA',
    label: 'Điện ảnh',
    description: 'Chữ thanh, bóng nhẹ và không dùng hộp nền.',
    style: {
      ...defaultVietsubSubtitleStyle,
      presetId: 'CINEMA',
      fontFamily: 'Times New Roman',
      fontSizePercent: 4.2,
      italic: true,
      backgroundEnabled: false,
      outlineWidthPercent: 0.22,
      shadowOffsetPercent: 0.2,
      bottomMarginPercent: 8,
      positionYPercent: 92
    }
  },
  {
    id: 'MINIMAL',
    label: 'Tối giản',
    description: 'Chữ gọn, viền rõ và không có nền.',
    style: {
      ...defaultVietsubSubtitleStyle,
      presetId: 'MINIMAL',
      fontFamily: 'Segoe UI',
      fontSizePercent: 3.8,
      backgroundEnabled: false,
      outlineWidthPercent: 0.2,
      shadowOffsetPercent: 0,
      bottomMarginPercent: 6,
      positionYPercent: 94
    }
  }
];

export function subtitleTextCss(
  style: VietsubSubtitleStyle,
  contentHeight: number
): CSSProperties {
  const safeHeight = Math.max(1, contentHeight);
  const outline = roundCssPixel(safeHeight * style.outlineWidthPercent / 100);
  const shadow = roundCssPixel(safeHeight * style.shadowOffsetPercent / 100);
  return {
    maxWidth: `${style.maxWidthPercent}%`,
    fontFamily: style.fontFamily,
    fontSize: `${roundCssPixel(Math.max(10, safeHeight * style.fontSizePercent / 100))}px`,
    fontWeight: style.bold ? 700 : 400,
    fontStyle: style.italic ? 'italic' : 'normal',
    lineHeight: style.lineHeight,
    color: colorWithOpacity(style.textColor, style.textOpacity),
    WebkitTextStroke: outline > 0
      ? `${outline}px ${colorWithOpacity(style.outlineColor, style.outlineOpacity)}`
      : undefined,
    paintOrder: 'stroke fill',
    textShadow: shadow > 0
      ? `${shadow}px ${shadow}px ${Math.max(0.5, shadow * 0.4)}px ${colorWithOpacity(style.shadowColor, style.shadowOpacity)}`
      : 'none',
    background: style.backgroundEnabled
      ? colorWithOpacity(style.backgroundColor, style.backgroundOpacity)
      : 'transparent',
    WebkitBoxOrient: 'vertical'
  };
}

export function hasSubtitleContrastWarning(style: VietsubSubtitleStyle): boolean {
  if (style.textOpacity < 0.7) return true;
  if (style.backgroundEnabled && style.backgroundOpacity >= 0.2) return false;
  return style.outlineWidthPercent < 0.12 || style.outlineOpacity < 0.45;
}

export function colorWithOpacity(hex: string, opacity: number): string {
  const normalized = /^#[0-9a-f]{6}$/i.test(hex) ? hex.slice(1) : '000000';
  const red = Number.parseInt(normalized.slice(0, 2), 16);
  const green = Number.parseInt(normalized.slice(2, 4), 16);
  const blue = Number.parseInt(normalized.slice(4, 6), 16);
  return `rgba(${red}, ${green}, ${blue}, ${Math.min(1, Math.max(0, opacity))})`;
}

export function cloneSubtitleStyle(style: VietsubSubtitleStyle): VietsubSubtitleStyle {
  return { ...style };
}

function roundCssPixel(value: number): number {
  return Math.round(value * 1000) / 1000;
}

File: TOOL-LOCAL/Vietsub/Domain/VietsubSubtitleStyle.cs
using System.Text.RegularExpressions;

namespace TOOL_LOCAL.Vietsub.Domain;

internal static class VietsubSubtitleStylePresetIds
{
    public const string Readable = "READABLE";
    public const string Vertical = "VERTICAL";
    public const string Outline = "OUTLINE";
    public const string TikTok = "TIKTOK";
    public const string Cinema = "CINEMA";
    public const string Yellow = "YELLOW";
    public const string Minimal = "MINIMAL";
    public const string Custom = "CUSTOM";

    public static readonly IReadOnlySet<string> Supported = new HashSet<string>(StringComparer.Ordinal)
    {
        Readable,
        Vertical,
        Outline,
        TikTok,
        Cinema,
        Yellow,
        Minimal,
        Custom
    };
}

internal static class VietsubSubtitleVerticalPositions
{
    public const string Top = "TOP";
    public const string Middle = "MIDDLE";
    public const string Bottom = "BOTTOM";
    public const string Custom = "CUSTOM";

    public static readonly IReadOnlySet<string> Supported = new HashSet<string>(StringComparer.Ordinal)
    {
        Top,
        Middle,
        Bottom,
        Custom
    };
}

internal static class VietsubSubtitleAlignments
{
    public const string BottomLeft = "BOTTOM_LEFT";
    public const string BottomCenter = "BOTTOM_CENTER";
    public const string BottomRight = "BOTTOM_RIGHT";

    public static readonly IReadOnlySet<string> Supported = new HashSet<string>(StringComparer.Ordinal)
    {
        BottomLeft,
        BottomCenter,
        BottomRight
    };
}

internal sealed partial class VietsubSubtitleStyle
{
    private static readonly IReadOnlyDictionary<string, string> SupportedFonts =
        new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["Arial"] = "Arial",
            ["Segoe UI"] = "Segoe UI",
            ["Tahoma"] = "Tahoma",
            ["Verdana"] = "Verdana",
            ["Times New Roman"] = "Times New Roman"
        };

    public string PresetId { get; set; } = VietsubSubtitleStylePresetIds.Readable;

    public string FontFamily { get; set; } = "Arial";

    public double FontSizePercent { get; set; } = 4.4;

    public bool Bold { get; set; } = true;

    public bool Italic { get; set; }

    public string TextColor { get; set; } = "#FFFFFF";

    public double TextOpacity { get; set; } = 1;

    public string OutlineColor { get; set; } = "#000000";

    public double OutlineOpacity { get; set; } = 0.95;

    public double OutlineWidthPercent { get; set; } = 0.18;

    public string ShadowColor { get; set; } = "#000000";

    public double ShadowOpacity { get; set; } = 0.65;

    public double ShadowOffsetPercent { get; set; } = 0.16;

    public bool BackgroundEnabled { get; set; } = true;

    public string BackgroundColor { get; set; } = "#04080D";

    public double BackgroundOpacity { get; set; } = 0.58;

    public string Alignment { get; set; } = VietsubSubtitleAlignments.BottomCenter;

    public string VerticalPosition { get; set; } = VietsubSubtitleVerticalPositions.Bottom;

    public double PositionXPercent { get; set; } = 50;

    public double PositionYPercent { get; set; } = 93;

    public double BottomMarginPercent { get; set; } = 7;

    public double HorizontalMarginPercent { get; set; } = 6;

    public double MaxWidthPercent { get; set; } = 88;

    public double LineHeight { get; set; } = 1.25;

    public int MaxLines { get; set; } = 2;

    public static VietsubSubtitleStyle CreateDefault() => new();

    public VietsubSubtitleStyle Copy() => new()
    {
        PresetId = PresetId,
        FontFamily = FontFamily,
        FontSizePercent = FontSizePercent,
        Bold = Bold,
        Italic = Italic,
        TextColor = TextColor,
        TextOpacity = TextOpacity,
        OutlineColor = OutlineColor,
        OutlineOpacity = OutlineOpacity,
        OutlineWidthPercent = OutlineWidthPercent,
        ShadowColor = ShadowColor,
        ShadowOpacity = ShadowOpacity,
        ShadowOffsetPercent = ShadowOffsetPercent,
        BackgroundEnabled = BackgroundEnabled,
        BackgroundColor = BackgroundColor,
        BackgroundOpacity = BackgroundOpacity,
        Alignment = Alignment,
        VerticalPosition = VerticalPosition,
        PositionXPercent = PositionXPercent,
        PositionYPercent = PositionYPercent,
        BottomMarginPercent = BottomMarginPercent,
        HorizontalMarginPercent = HorizontalMarginPercent,
        MaxWidthPercent = MaxWidthPercent,
        LineHeight = LineHeight,
        MaxLines = MaxLines
    };

    public void Normalize()
    {
        PresetId = (PresetId ?? string.Empty).Trim().ToUpperInvariant();
        if (!VietsubSubtitleStylePresetIds.Supported.Contains(PresetId))
        {
            throw new ArgumentException("Mẫu thiết kế phụ đề không hợp lệ.", nameof(PresetId));
        }

        var requestedFont = (FontFamily ?? string.Empty).Trim();
        if (!SupportedFonts.TryGetValue(requestedFont, out var safeFont))
        {
            throw new ArgumentException("Phông chữ phụ đề không nằm trong danh sách được hỗ trợ.", nameof(FontFamily));
        }
        FontFamily = safeFont;

        Alignment = (Alignment ?? string.Empty).Trim().ToUpperInvariant();
        if (!VietsubSubtitleAlignments.Supported.Contains(Alignment))
        {
            throw new ArgumentException("Vị trí phụ đề không hợp lệ.", nameof(Alignment));
        }


        VerticalPosition = (VerticalPosition ?? string.Empty).Trim().ToUpperInvariant();
        if (!VietsubSubtitleVerticalPositions.Supported.Contains(VerticalPosition))
        {
            throw new ArgumentException("Vị trí dọc của phụ đề không hợp lệ.", nameof(VerticalPosition));
        }

        TextColor = NormalizeColor(TextColor, nameof(TextColor));
        OutlineColor = NormalizeColor(OutlineColor, nameof(OutlineColor));
        ShadowColor = NormalizeColor(ShadowColor, nameof(ShadowColor));
        BackgroundColor = NormalizeColor(BackgroundColor, nameof(BackgroundColor));

        ValidateRange(FontSizePercent, 2, 10, nameof(FontSizePercent));
        ValidateRange(TextOpacity, 0, 1, nameof(TextOpacity));
        ValidateRange(OutlineOpacity, 0, 1, nameof(OutlineOpacity));
        ValidateRange(OutlineWidthPercent, 0, 1, nameof(OutlineWidthPercent));
        ValidateRange(ShadowOpacity, 0, 1, nameof(ShadowOpacity));
        ValidateRange(ShadowOffsetPercent, 0, 1.5, nameof(ShadowOffsetPercent));
        ValidateRange(BackgroundOpacity, 0, 1, nameof(BackgroundOpacity));
        ValidateRange(PositionXPercent, 2, 98, nameof(PositionXPercent));
        ValidateRange(PositionYPercent, 2, 98, nameof(PositionYPercent));
        ValidateRange(BottomMarginPercent, 0, 30, nameof(BottomMarginPercent));
        ValidateRange(HorizontalMarginPercent, 0, 30, nameof(HorizontalMarginPercent));
        ValidateRange(MaxWidthPercent, 40, 100, nameof(MaxWidthPercent));
        ValidateRange(LineHeight, 1, 2, nameof(LineHeight));
        if (MaxLines is < 1 or > 3)
        {
            throw new ArgumentOutOfRangeException(nameof(MaxLines), "Số dòng phụ đề mục tiêu phải từ 1 đến 3.");
        }

        if ((HorizontalMarginPercent * 2) + MaxWidthPercent > 100.001)
        {
            throw new ArgumentException(
                "Lề ngang và chiều rộng tối đa của phụ đề vượt quá khung hình.",
                nameof(MaxWidthPercent));
        }
    }

    private static string NormalizeColor(string? value, string parameterName)
    {
        var normalized = (value ?? string.Empty).Trim().ToUpperInvariant();
        if (!HexColorPattern().IsMatch(normalized))
        {
            throw new ArgumentException("Màu phụ đề phải có định dạng #RRGGBB.", parameterName);
        }
        return normalized;
    }

    private static void ValidateRange(double value, double minimum, double maximum, string parameterName)
    {
        if (!double.IsFinite(value) || value < minimum || value > maximum)
        {
            throw new ArgumentOutOfRangeException(
                parameterName,
                $"Giá trị thiết kế phụ đề phải nằm trong khoảng {minimum} đến {maximum}.");
        }
    }

    [GeneratedRegex("^#[0-9A-F]{6}$", RegexOptions.CultureInvariant)]
    private static partial Regex HexColorPattern();
}
